### PR TITLE
Fixes #18304 - Support network edit for vmware vm

### PR DESCRIPTION
**Pull request by leewaa**
Just waiting on some in-coming changes to `vog-vsphere`.

### CODE REVIEW

**Comment by leewaa** [comment]
OK basically done here, the PR fog-vsphere is https://github.com/fog/fog-vsphere/pull/142 there is still a little bit of work to be done there so will wait on that before setting it free to merge. The PR here is good for a review of code and any implications it could have.

Also take note: adding new interfaces does not currently work, and since I think it has never worked, I will take care of this in a separate issue since it does expand the scope somewhat of the original issue.

File: app/controllers/hosts_controller.rb
Comment (timogoebel) [inline app/controllers/hosts_controller.rb:116]: Do we need this in APIv2 as well?

**Review by timogoebel — COMMENTED** [review]
Nice, looks good to me. I left one comment inline.

File: app/controllers/hosts_controller.rb
Comment (leewaa) [reply app/controllers/hosts_controller.rb:116]: Not sure about this. Maybe someone else can provide more insight ?

Comment (timogoebel) [reply app/controllers/hosts_controller.rb:116]: We do. :-)

**Comment by leewaa** [comment]
Ok this now includes the addition to the API also. I rebased against develop to resolve some merge conflicts. Should be good to go now.

File: app/controllers/api/v2/hosts_controller.rb
Comment (mmoll) [inline app/controllers/api/v2/hosts_controller.rb]: is this the same `if host` as in the line above? :)

Comment (leewaa) [reply app/controllers/api/v2/hosts_controller.rb]: oh I thought I already removed this, good catch, thanks :D

Comment (leewaa) [reply app/controllers/api/v2/hosts_controller.rb]: Fixed!

**Comment by mmoll** [comment]
@leewaa please rebase :)

**Comment by leewaa** [comment]
Can we get this merged ?  the tests seem to be unrelated

**Comment by timogoebel** [comment]
> Can we get this merged

@leewaa: I'll do a (final) round of review tomorrow.

File: app/models/host/managed.rb
Comment (timogoebel) [inline app/models/host/managed.rb]: I don't like that we match networks in Host::Managed as this might be compute_resource specific. I have an implementation ready where this is not required: https://github.com/timogoebel/foreman/commit/f672034e42f4c2246aa8d20aaed4e4a128e1eaae

Do you mind taking a look and updating this PR accordingly? Thanks.

**Review by timogoebel — CHANGES_REQUESTED** [review]
@leewaa: Thanks, I left a comment inline with some (working) example code.

File: app/models/host/managed.rb
Comment (leewaa) [reply app/models/host/managed.rb]: There is a small, but important, issue with this requested change.

The `compute_attributes` are used to determine whether a compute resource needs updating or not. See: `compute_update_required?` https://github.com/theforeman/foreman/blob/8397b3c93489ecf31d2b8cf88c50439709e29d50/app/models/concerns/orchestration/compute.rb#L232

Where `update_requried?` on the compares the key-values for changes. (https://github.com/theforeman/foreman/blob/develop/app/models/compute_resource.rb#L228) With this change it will always trigger a compute resource update since the network value will be something like 123456 (comes from the form) and not the actual name of the network.

An idea would be to change the form's select field to use the network's name as the value instead of the network id

Comment (timogoebel) [reply app/models/host/managed.rb]: I changed the format of the compute attributes to always show the id and not the name. So that form and attributes are the same if nothing changes.

**Comment by leewaa** [comment]
Just revisited this. Is there any reason this was left hanging around ?

After thoroughly testing it the feature seems to be working perfectly with this implementation.

@lzap maybe you can take a look at this ? would be nice to get this one merged soon, it has been open since May.

**Comment by lzap** [comment]
I don't have a VMWare setup, @timogoebel is our guy for these kind of PRs. Code-wise I have no comments.

**Comment by leewaa** [comment]
@lzap the endpoint should be working for bmc machines also, so feel free to test :)

**Comment by lzap** [comment]
Tested and merged the BMC endpoint that was the other PR I suppose, this one is pure VMWare I guess?

**Comment by timogoebel** [comment]
> this one is pure VMWare I guess?

Yeah. I hope I get to this on Friday. It's crazy, I barely have time to review and test PRs.

**Comment by dgoetz** [comment]
@leewaa @timogoebel: 
I tested it but had the following issues:
- When you open the interface edit form, it shows the wrong VLAN (always the first of the selection) until you stored it once after applying the patch.
- The list of VLANs contains all VLANs and not only those useable for the VM. Choosing a non useable and submitting does result in a success message, but changes only the VM in Foreman but not in VMware.
If you select correct ones, it works fine. 

So I would say if both issues are easily fixed, we should do so before merging. If the fix for the second one is not easy, I would be good with merging and solving it in a follow-up PR.

**Comment by timogoebel** [comment]
> * When you open the interface edit form, it shows the wrong VLAN (always the first of the selection) until you stored it once after applying the patch.

I think the network attributes are stored/cached in the database (as a serialized attribute). I guess Foreman should show the actual data from the VM instead (it feches it anyway, all the information in the Virtual Maschine tab is fetched via the compute ressources' api).

**Comment by leewaa** [comment]
Thanks @dgoetz and @timogoebel will get adjust a per your comments, happy to get this one finished off.

**Comment by leewaa** [comment]
Just tested this again with @dgoetz and all seems to be working as expected. @dgoetz is not sure how he was able to produce what he described above in the first point of his comment... We tested this several times and it is working as expected. It will also fail to update the vm if the network change is not a valid one.

As for the second point, this is also the case for Host creation, here an error message is displayed and the host cannot be created, just as edit. I would suggest filtering the list of selectable vlans be done in a separate PR.

For what @timogoebel  suggested I'm not totally against it however I advise against it. Since it would mean another request for fetching this info, in turn slowing down the edit view further. We were not able to produce any problems here, and it is not the only area that can become out of sync when vms are modified directly in VMware and co outside of foreman. WE can keep an eye on it and if it becomes a real problem then I would suggest changing it in a separate PR.

**Comment by leewaa** [comment]
I just fixed that failing test. Not sure about the others that were red, but will wait for the build to finish then re-asses.

**Comment by leewaa** [comment]
Nice feature with the foreman bot about rebasing 👍

**Comment by leewaa** [comment]
Is this waiting on anything @timogoebel @lzap

File: app/controllers/api/v2/hosts_controller.rb
Comment (lzap) [inline app/controllers/api/v2/hosts_controller.rb:361]: Is this intentional? You assign to variable and then re-assign it once again, then return it?

File: app/models/concerns/orchestration/compute.rb
Comment (lzap) [inline app/models/concerns/orchestration/compute.rb:72]: This should be really a debug message as this can be often the case.

**Review by lzap — COMMENTED** [review]
Code-wise fine from me, one small remark. But again I don't test VMWare, maybe @chris1984 or @timogoebel could test this more quickly?

**Comment by chris1984** [comment]
@lzap I will start testing it, I have been working on testing prs all week so we can get these all wrapped up.

File: app/models/compute_resources/foreman/model/vmware.rb
Comment (dgoetz) [inline app/models/compute_resources/foreman/model/vmware.rb:480]: `attr` has to be changed to `attrs`, too.

File: app/models/concerns/orchestration/compute.rb
Comment (dgoetz) [reply app/models/concerns/orchestration/compute.rb:72]: Works fine also with `logger.debug`.

**Review by dgoetz — CHANGES_REQUESTED** [review]
I retested it with the changed log level @lzap suggested and found a small code change required see inline comments.
And I still see the problem from my first test: When you open the interface edit form, it shows the wrong VLAN (always the first of the selection) until you stored it once after applying the patch. Is this solvable to not confuse people?

**Comment by chris1984** [comment]
Will hold off on testing until @dgoetz comments have been addressed so we can test with everything correct.

**Comment by timogoebel** [comment]
@ezr-ondrej: Do you want to take ownership of this PR and finish it? @leewaa is not going to finish this anymore.

**Comment by ezr-ondrej** [comment]
ok, if @leewaa don't mind, I will add it to my backlog :)

**Comment by leewaa** [comment]
Go for it. Since I'm not actively working on theforeman anymore feel free to take it over.

**Comment by ares** [comment]
Thanks @leewaa, this has been taken over by @ezr-ondrej who has the WIP private branch and hopefully will open a PR soon. We're closing this PR for now.

**Comment by Fodoj** [comment]
Is it still in progress? I could not find a follow up take over PR

**Comment by wiad** [comment]
@ezr-ondrej This would be really great for us, is there any plans to move forward with this?

**Comment by wiad** [comment]
I have just tested this code on Foreman 2.0.0 and it works like a charm. The issue that the wrong VLAN shows in the interface edit box is however present for me as well, but that is the case without the patch as well. In my case I believe this is because the vlan configuration in my compute profiles got reset so all compute profiles had the wrong vlan - the one now showing in the edit interface. So it seems to show not what is actually set in VMware but rather what was set in the compute profile when the host was created. All hosts that have been installed after I fixed my compute profiles shows the correct vlan in the edit interface, both with and without this patch.

**Comment by wiad** [comment]
Can't get it to work via API though. Seems like the each loop in `apply_interfaces_compute_attributes` function does not work with the attributes hash provided by the api.

**Comment by ezr-ondrej** [comment]
Yeah, this is not missing much, I've got a branch and the `fog-vsphere` changes are in place, I can open a PR and if you'd be able to test it, it would be a great help.

**Comment by chris1984** [comment]
@ezr-ondrej do we need Foreman testing on this?

**Comment by ezr-ondrej** [comment]
Yeah, once I'll open the PR, that would be nice :)
I'll ping you there

**Comment by wiad** [comment]
@ezr-ondrej is there a PR for this now that I've missed? I would be happy to help with testing.

**Comment by ezr-ondrej** [comment]
Hi, sorry for not updating this, I've only now realized :facepalm:
Here is a new PR: #7991 if anyone could test  it, that would be great :)

**Pull request closed without merging**
